src/sentry/web/frontend/oauth_token.py

```diff
@@ -41,85 +43,103 @@ def error(self, request: HttpRequest, name, reason=None, status=400):
-            except ApiToken.DoesNotExist:
-                return self.error(request, "invalid_grant", "invalid token")
-
-            token.refresh()
+            return self._get_access_tokens(request, client_id)
+        elif grant_type == GrantTypes.REFRESH:
+            return self._get_refresh_token(request, client_id)
         else:
             return self.error(request, "unsupported_grant_type")
 
+    def _get_access_tokens(self, request, client_id):
+        redirect_uri = request.POST.get("redirect_uri")
+        code = request.POST.get("code")
+
+        try:
+            application = ApiApplication.objects.get(
+                client_id=client_id, status=ApiApplicationStatus.active
+            )
+        except ApiApplication.DoesNotExist:
+            return self.error(request, "invalid_client", "invalid client_id")
+
+        try:
+            grant = ApiGrant.objects.get(application=application, code=code)
+        except ApiGrant.DoesNotExist:
+            return self.error(request, "invalid_grant", "invalid grant")
+
+        if grant.is_expired():
+            return self.error(request, "invalid_grant", "grant expired")
+
+        if not redirect_uri:
+            redirect_uri = application.get_default_redirect_uri()
+        elif grant.redirect_uri != redirect_uri:
+            return self.error(request, "invalid_grant", "invalid redirect_uri")
+
+        access_token = ApiToken.from_grant(grant)
+        id_token = self._get_open_id_token(grant, request)
+        return self._process_token_details(access_token, id_token)
+
+    def _get_open_id_token(self, grant, request):
+        if grant.has_scope("openid"):
+            open_id_token = OpenIDToken(
+                request.POST.get("client_id"),
+                grant.user.id,
+                # Encrypt with a random secret until we implement secure shared secrets in prod
+                secrets.token_urlsafe(),
+                nonce=request.POST.get("nonce"),
+            )
+            return open_id_token.get_encrypted_id_token()
+        return None
+
+    def _get_refresh_token(self, request, client_id):
+        refresh_token = request.POST.get("refresh_token")
+        scope = request.POST.get("scope")
+
+        if not refresh_token:
+            return self.error(request, "invalid_request")
+
+        # TODO(dcramer): support scope
+        if scope:
+            return self.error(request, "invalid_request")
+
+        try:
+            application = ApiApplication.objects.get(
+                client_id=client_id, status=ApiApplicationStatus.active
+            )
+        except ApiApplication.DoesNotExist:
+            return self.error(request, "invalid_client", "invalid client_id")
+
+        try:
+            token = ApiToken.objects.get(application=application, refresh_token=refresh_token)
+        except ApiToken.DoesNotExist:
+            return self.error(request, "invalid_grant", "invalid token")
```

Comment:
```suggestion
            return self.error(request, "invalid_token", "invalid refresh token")
```

This error message should reflect the problem.